typebackend/typingviews.py

```diff
@@ -3,26 +3,64 @@
 from django.contrib.auth.models import User
-from .models import PractiseLog,Paragraph
-from .serializers import PractiseLogSerializer,ParagraphSerializer
+from .models import PractiseLog,Paragraph,DashboardData
+from .serializers import PractiseLogSerializer,ParagraphSerializer,StreakSerializer
 from rest_framework.permissions import IsAuthenticated,IsAdminUser
 from rest_framework.authtoken.models import Token
+from django.db.models import Sum
 from django.utils import timezone
 from random import randint,choice
 import datetime
 import json
 
+def create_or_update_streak(user,new_entry,mode): 
+    # Converting string date to datetime format
+    new_entry=datetime.datetime.strptime(new_entry,'%Y-%m-%d %H:%M:%S')
+
+    try:
+        last_typed=PractiseLog.objects.filter(user=user).latest('taken_at').taken_at
+    except PractiseLog.DoesNotExist:
+        # When its a new user, the first if condition has to be followed. Hence setting last_typed to previous_day
+        last_typed=new_entry-datetime.timedelta(1)
+
+    # Finding the difference in days between the last log and the new log.
+    days_between_recent_and_lastlog=(new_entry-last_typed).days
+    data=DashboardData.objects.get(user=user)
+
+    # Getting the existing value of the mode [practise/arcade/race] specified in the user object from the query and updating it.
+    count=getattr(data,mode)+1
+    #Setting the updated value of the mode count
+    setattr(data,mode,count)
+
+    # Checking for consecutive streak from the database, i.e if the difference between previous log and new log is 1, then the user has been typing consecutively
+    if days_between_recent_and_lastlog==1:
+        data.streak=data.streak+1
+        data.total_streak=1 if data.total_streak==0 else 0 # Incase of new user
+        data.longest_streak=data.streak if data.streak>data.longest_streak else data.longest_streak
+    # If the difference is greater, then reset the counter
+    if days_between_recent_and_lastlog>1:
+        data.inactive_days=data.inactive_days+(days_between_recent_and_lastlog-1)
+        data.streak=1
+        data.total_streak=data.total_streak+1
+
+    data.save()
+
 class PostSpeed(APIView):
     permission_classes=[IsAuthenticated]
 
     def post(self,request):
-        serializers=PractiseLogSerializer(data=request.data,context={'request':request})
-        
-        if serializers.is_valid():  
-            user=serializers.save()
-            return Response({'success':True})
-        else:
+        if request.data['mode']=='practise' or request.data['mode']=='race' or request.data['mode']=='arcade':
+            serializers=PractiseLogSerializer(data=request.data,context={'request':request})
+            if serializers.is_valid():
+                user=request.user
+                typed_at=request.data['taken_at']
+                mode=request.data['mode']
+                # Streak counter update 
+                create_or_update_streak(user,typed_at,mode) 
```

Comment:
Actually this whole `create_or_update_streak` can be moved as a Django signal.
Read about on_create signal. It doesn't block the request thread.